File: packages/camera/camera_avfoundation/example/ios/RunnerTests/Mocks/MockCaptureSession.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "MockCaptureSession.h"

@implementation MockCaptureSession

- (instancetype)init {
  self = [super init];
  if (self) {
    _inputs = [NSMutableArray array];
    _outputs = [NSMutableArray array];
  }
  return self;
}

- (void)beginConfiguration {
  if (self.beginConfigurationStub) {
    self.beginConfigurationStub();
  }
}

- (void)commitConfiguration {
  if (self.commitConfigurationStub) {
    self.commitConfigurationStub();
  }
}

- (void)startRunning {
  if (self.startRunningStub) {
    self.startRunningStub();
  }
}

- (void)stopRunning {
  if (self.stopRunningStub) {
    self.stopRunningStub();
  }
}

- (BOOL)canSetSessionPreset:(AVCaptureSessionPreset)preset {
  return self.canSetSessionPreset;
}

- (void)addConnection:(nonnull AVCaptureConnection *)connection {
}

- (void)addInput:(nonnull NSObject<FLTCaptureInput> *)input {
}

- (void)addInputWithNoConnections:(nonnull NSObject<FLTCaptureInput> *)input {
}

- (void)addOutput:(nonnull AVCaptureOutput *)output {
}

- (void)addOutputWithNoConnections:(nonnull AVCaptureOutput *)output {
}

- (BOOL)canAddConnection:(nonnull AVCaptureConnection *)connection {
  return YES;
}

- (BOOL)canAddInput:(nonnull NSObject<FLTCaptureInput> *)input {
  return YES;
}

- (BOOL)canAddOutput:(nonnull AVCaptureOutput *)output {
  return YES;
}

- (void)removeInput:(nonnull NSObject<FLTCaptureInput> *)input {
}

- (void)removeOutput:(nonnull AVCaptureOutput *)output {
}

- (void)setSessionPreset:(AVCaptureSessionPreset)sessionPreset {
  if (self.setSessionPresetStub) {
    self.setSessionPresetStub(sessionPreset);
  }
}

@end

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/Mocks/MockCaptureSession.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

@testable import camera_avfoundation

// Import Objective-C part of the implementation when SwiftPM is used.
#if canImport(camera_avfoundation_objc)
  import camera_avfoundation_objc
#endif

/// Mock implementation of `FLTCaptureSession` protocol which allows injecting a custom
/// implementation.
final class MockCaptureSession: NSObject, CaptureSession {
  var setSessionPresetStub: ((AVCaptureSession.Preset) -> Void)?
  var beginConfigurationStub: (() -> Void)?
  var commitConfigurationStub: (() -> Void)?
  var startRunningStub: (() -> Void)?
  var stopRunningStub: (() -> Void)?
  var canSetSessionPresetStub: ((AVCaptureSession.Preset) -> Bool)?

  var _sessionPreset = AVCaptureSession.Preset.high
  var inputs = [AVCaptureInput]()
  var outputs = [AVCaptureOutput]()

  private(set) var addedAudioOutputCount: Int = 0

  var automaticallyConfiguresApplicationAudioSession = false

  var sessionPreset: AVCaptureSession.Preset {
    get {
      return _sessionPreset
    }
    set {
      setSessionPresetStub?(newValue)
    }
  }

  func beginConfiguration() {
    beginConfigurationStub?()
  }

  func commitConfiguration() {
    commitConfigurationStub?()
  }

  func startRunning() {
    startRunningStub?()
  }

  func stopRunning() {
    stopRunningStub?()
  }

  func canSetSessionPreset(_ preset: AVCaptureSession.Preset) -> Bool {
    return canSetSessionPresetStub?(preset) ?? true
  }

  func addInputWithNoConnections(_ input: CaptureInput) {}

  func addOutputWithNoConnections(_ output: AVCaptureOutput) {}

  func addConnection(_: AVCaptureConnection) {}

  func addInput(_: CaptureInput) {}

  func addOutput(_ output: AVCaptureOutput) {

    if output is AVCaptureAudioDataOutput {
      addedAudioOutputCount += 1
    }
  }

  func removeInput(_: CaptureInput) {}

  func removeOutput(_: AVCaptureOutput) {}

  func canAddInput(_: CaptureInput) -> Bool {
    return true
  }

  func canAddOutput(_: AVCaptureOutput) -> Bool {
    return true
  }

  func canAddConnection(_: AVCaptureConnection) -> Bool {
    return true
  }
}
